### PR TITLE
Added a unit test for the iOS AccessibilityBridge.

### PR DESCRIPTION
**Pull request by gaaclarke**
This is setting up the infrastructure to test the iOS AccessibilityBridge.

Related PR: https://github.com/flutter/engine/pull/18164

### CODE REVIEW

**Comment by gaaclarke** [comment]
I threw a few extra tests onto this PR while I was waiting for a review since github doesn't handle dependent PRs very well.  I'll refrain from adding more.

File: fml/platform/darwin/scoped_nsobject.h [hunk 1]
```diff
@@ -42,15 +42,37 @@ class scoped_nsprotocol {
  public:
   explicit scoped_nsprotocol(NST object = nil) : object_(object) {}
 
-  scoped_nsprotocol(const scoped_nsprotocol<NST>& that) : object_([that.object_ retain]) {}
+  scoped_nsprotocol(const scoped_nsprotocol<NST>& that)
```
Comment (chinmaygarde) [inline fml/platform/darwin/scoped_nsobject.h:45]: Handling of inclusion from arc and non-arc TUs is unrelated to AccessibilityBridge. Can you separate this out into its own patch and add a unit-test (in the fml_unittests harness) that ensures objects are correctly released. Crucially, I am not able to fully reason about the correctness of what happens when say an ARC TU vends an API to a non-ARC TU using this header and vice-versa.

cc @stuartmorgan who has experience with ARC migrations.

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm [hunk 2]
```diff
@@ -0,0 +1,130 @@
+#import <XCTest/XCTest.h>
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm:1]: Standard Flutter license verbiage.

Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm:5]: Done.

File: fml/platform/darwin/scoped_nsobject.h (hunk 1 above)
Comment (gaaclarke) [reply fml/platform/darwin/scoped_nsobject.h:45]: Unfortunately it is not unrelated, it is necessary for this test to function.  Without it code that includes accessibility_bridge.h won't compile under ARC and OCMock breaks without ARC.

I could refactor the bridge to use the pimpl pattern but I think the right move is just to fix scoped_nsobject so we fix it everywhere once.

I can add a test that shows this works.

Comment (chinmaygarde) [reply fml/platform/darwin/scoped_nsobject.h:45]: > Unfortunately it is not unrelated, it is necessary for this test to function.

Can you clarify? I am not sure why the test cannot use manual reference counting in the test and compile the OCMock TUs with ARC. I'd like to understand why that causes an issue because I am afraid of introducing exactly the same kinds of issues here in the engine.

Comment (gaaclarke) [reply fml/platform/darwin/scoped_nsobject.h:45]: Sure.  In my experience OCMock has problems in manually reference counted code.  I tried to find reference to the problems but couldn't pull anything up.  What it comes down to is that ARC has been adopted by the majority of developers and OCMock is doing funky things to work.  It just isn't as tested for manual reference counting.  I wouldn't consider it a concern for ARC adoption.  It is only an issue because of the odd nature of mocking libraries.

This change doesn't introduce ARC to the engine.  The engine will execute exactly the same.  It just allows us to continue writing our unit tests in ARC.  You'll have to take my word that OCMock works better with ARC.  This PR just makes our code ARC compatible, which is a positive move.

Comment (gaaclarke) [reply fml/platform/darwin/scoped_nsobject.h:45]: Since there is still concern I added a unit-test.

**Comment by gaaclarke** [comment]
This change to scoped_nsobject is breaking scenario_app tests.  I'm a bit baffled why.  They are using ARC but never including scoped_nsobject directly so I would have expected an identical binary to be produced.  I'll go back to the drawing board.

**Comment by gaaclarke** [comment]
@chinmaygarde Okay, I rewrote it and I think you'll find this more agreeable.  Now `scoped_nsobject` can be see by ARC code, but it can't be used by ARC maintained objects (that will result in a linking error).  The change is very straightforward and the scenario_app and unit tests pass locally.

**Comment by chinmaygarde** [comment]
Taking a closer look. I think you will agree that the scope of this patch expands well beyond just the iOS accessibility bridge. Unless technically unfeasible, please use MRC in the unit-test and lets land this ASAP.

**Comment by gaaclarke** [comment]
@chinmaygarde Thanks.  I think you'll be OK with this change.

It's not that MRC is technically impossible (it might be depending on how OCMock is being used).  Transitioning to MRC is more work and we end up in an inferior position with respect to testing stability, consistency, convenience and compatibility.  All of our iOS unit tests are ARC.  It will be confusing for contributors to have some tests written one way and some tests another.  Even if OCMock works with the subset of usages I'm using today, someone may have to rewrite it in ARC when they want to use a particular feature of OCMock.

**Comment by gaaclarke** [comment]
@jmagman Hey Jenn, can you take a look at this when you get a chance.  @chinmaygarde and I would like you to double check this, thanks!

File: fml/platform/darwin/scoped_nsobject.h [hunk 3]
```diff
@@ -37,20 +37,26 @@ namespace fml {
 // We check for bad uses of scoped_nsobject and NSAutoreleasePool at compile
 // time with a template specialization (see below).
 
+id ObjcRetain(id object);
+
+id ObjcAutorelease(id object);
+
+void ObjcRelease(id object);
```
Comment (jmagman) [inline fml/platform/darwin/scoped_nsobject.h:44]: Can you explain why these exist?

Comment (gaaclarke) [reply fml/platform/darwin/scoped_nsobject.h:44]: Sure.  It's a bit of an oddity, a collision between how C++ macros work and ARC.  Get ready for a long explanation.

The goal of this PR is to test the C++ class `flutter::AccessibilityBridge`, that class has instance variables of type `fml::scoped_nsobject`.  The way C++ works is that it must declare its instance variables in the header so that the size of the object is known.  `fml::scoped_nsobject` is a template class which will compile for each compilation unit the instances that it finds are actually used, when it is used.  They aren't precompiled and may actually result in no code getting generated if they are never used.

So, `fml::scoped_nsobject` is just a RAII wrapper around objective-c manual reference counting.  The problem is that when we compile accessibility_bridge_test.mm (which uses ARC like all our tests), it includes accessibility_bridge.h, which includes scoped_nsobject.h, which uses manual reference counting methods.  This creates an error when compiling `accessibility_bridge_test.mm` because it is compiled under ARC and can't deal with manual reference counting methods.

So, this PR wraps up the manual reference counting methods in C functions whose declaration is known to clients and users of the `fml::scoped_nsobject`, but whose actual definition is only known to the fml library, effectively hiding the forbidden methods and guaranteeing that the `fml::scoped_nsobject` will behave the same in an ARC or non-ARC compilation unit.

Long story short, If you remove these functions the test won't compile.

If you try to use an ARC managed Objective-C class with `fml::scoped_nsobject` you get a linker error because it won't be able to find the `retain` and `release` methods on that class.  This is how my second approach is different than my first, I tried to make fml::scoped_nsobject work for both type of objects (ARC and non-ARC) which didn't work out.

File: shell/platform/darwin/ios/framework/Source/SemanticsObject.h [hunk 4]
```diff
@@ -155,4 +155,48 @@ constexpr int32_t kRootNodeId = 0;
...
+ * world (second tree) as follows: We replace each SemanticsObjects that has children with a
+ * SemanticsObjectContainer, which has the original SemanticsObject and its children as children.
+ *
+ * SemanticsObjects have semantic information attached to them which is interpreted by
+ * VoiceOver (they return YES for isAccessibilityElement). The SemanticsObjectContainers are just
+ * there for structure and they don't provide any semantic information to VoiceOver (they return
+ * NO for isAccessibilityElement).
+ */
+@interface SemanticsObjectContainer : UIAccessibilityElement
+- (instancetype)init __attribute__((unavailable("Use initWithSemanticsObject instead")));
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/SemanticsObject.h]: Done.

File: fml/platform/darwin/scoped_nsobject.h (hunk 1 above)
Comment (stuartmorgan) [reply fml/platform/darwin/scoped_nsobject.h]: Have you looked at Chromium's current `scoped_ns*` implementations? Getting it right for cross-mode TUs was a nightmare, involving several iterations that turned out to have subtle undefined behavior bugs (once due to a non-obvious ODR violation, and once due to what eventually turned out to be a compiler bug IIRC).

Comment (gaaclarke) [reply fml/platform/darwin/scoped_nsobject.h]: https://chromium.googlesource.com/chromium/src/base/+/master/mac/scoped_nsobject.h#49

Thanks Stuart, they use an extra template parameter which allows you to supply retain and release implementations (presumably they pass in noops in ARC code).

This PR is slightly different.  I first tried to to make scoped_nsobject work for ARC and non-ARC instances and ran into weirdness.  I suspect it was due to ARC's usage of the autorelease pool so that deallocation time isn't deterministic.

This doesn't have that problem because it just wraps objc message calls in c functions.  I didn't try to make scoped_nsobject usable with ARC instances.  In fact you get a linker error if you try to do that.  I'm only interested in letting ARC instances use objects that have scoped_nsobjects.  If I'm writing ARC C++ code I won't need scoped_nsobject, I can just rely on ARC.  The fact that this works is exercised in the new unit test I added.

As far as concerns about weirdness, this just wraps objc messages in c functions, otherwise behaves the same.  The PR passes integration tests and unit tests.  Since the only ARC code is unit tests, the worse case scenario is that our tests will flake or crash.

Comment (stuartmorgan) [reply fml/platform/darwin/scoped_nsobject.h]: > I'm only interested in letting ARC instances use objects that have scoped_nsobjects.

That's what we were doing too, and we hit UB trying.

Comment (gaaclarke) [reply fml/platform/darwin/scoped_nsobject.h]: Do you have any documentation on what didn't work previously so we can see exactly what was tried and what went wrong?  Right now we have reason and empirical evidence that this works versus anecdotal evidence that something like it didn't work in a different context.

I appreciate the perspective.  The stakes are low here since this only affects testing targets.  Let's not hold back progress for a hypothetical, let's try to pin down the actual problem if one exists.

**Comment by gaaclarke** [comment]
At @cbracken 's request I split out https://github.com/flutter/engine/pull/18362.

File: shell/platform/darwin/ios/framework/Source/SemanticsObject.h [hunk 5]
```diff
@@ -155,4 +155,50 @@ constexpr int32_t kRootNodeId = 0;
...
+ * From Flutter's view of the world (the first tree seen above), we construct iOS's view of the
+ * world (second tree) as follows: We replace each SemanticsObjects that has children with a
+ * SemanticsObjectContainer, which has the original SemanticsObject and its children as children.
+ *
+ * SemanticsObjects have semantic information attached to them which is interpreted by
+ * VoiceOver (they return YES for isAccessibilityElement). The SemanticsObjectContainers are just
+ * there for structure and they don't provide any semantic information to VoiceOver (they return
+ * NO for isAccessibilityElement).
+ */
+@interface SemanticsObjectContainer : UIAccessibilityElement
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/SemanticsObject.h:192]: Any way we can start a policy of prefixing all classes with the `FLT` prefix (since this one has never been in a header before it's eligible even though it's not public)?  It's a poor namespacing substitute but it's all Obj-C's got.  We have random plugins using that prefix too, which is also not the right thing to do across frameworks...  The other classes at least start with `Flutter` which is still not good, but is at least some kind of prefix.
http://google.github.io/styleguide/objcguide.html#prefixes
https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/Conventions/Conventions.html#//apple_ref/doc/uid/TP40011210-CH10-SW4

**Comment by gaaclarke** [comment]
@chinmaygarde I underestimated the time it would take to get ARC support approval.  I've disentangled it for now.  As you can see this confuses the testing build target a bit but we can remove that once we land ARC support.  Please give this a glance.
